icepyx/core/read.py

```diff
@@ -367,19 +392,56 @@ def __init__(
             assert pattern_ck
             self._filelist = filelist
         elif isinstance(data_source, list):
+            # if data_source is a list pass that directly to _filelist
             self._filelist = data_source
         elif os.path.isdir(data_source):
+            # if data_source is a directory glob search the directory and assign to _filelist
             data_source = os.path.join(data_source, "*")
             self._filelist = glob.glob(data_source, **glob_kwargs)
+        elif isinstance(data_source, str):
+            if data_source.startswith("s3"):
+                # if the string is an s3 path put it in the _filelist without globbing
+                self._filelist = [data_source]
+            else:
+                # data_source is a globable string
+                self._filelist = glob.glob(data_source, **glob_kwargs)
         else:
-            self._filelist = glob.glob(data_source, **glob_kwargs)
-        # Remove any directories from the list
+            raise TypeError(
+                "data_source should be a list of files, a directory, the path to a file, "
+                "or a glob string."
+            )
+        # Remove any directories from the list (these get generated during recursive
+        # glob search)
         self._filelist = [f for f in self._filelist if not os.path.isdir(f)]
 
         # Create a dictionary of the products as read from the metadata
         product_dict = {}
-        for file_ in self._filelist:
-            product_dict[file_] = is2ref.extract_product(file_)
+        self.is_s3 = [False] * len(self._filelist)
+        for i, file_ in enumerate(self._filelist):
+            # If the path is an s3 path set the respective element of self.is_s3 to True
+            if file_.startswith("s3"):
+                self.is_s3[i] = True
+                auth = self.auth
+            else:
+                auth = None
+            product_dict[file_] = is2ref.extract_product(file_, auth=auth)
+
+        # Raise an error if there are both s3 and non-s3 paths present
+        if len(set(self.is_s3)) > 1:
+            raise TypeError(
+                "Mixed local and s3 paths is not supported. data_source must contain "
+                "only s3 paths or only local paths"
+            )
+        self.is_s3 = self.is_s3[0]  # Change is_s3 into one boolean value for _filelist
+        # Raise warning if more than 2 s3 files are given
+        if self.is_s3 is True and len(self._filelist) > 2:
+            warnings.warn(
+                "Processing more than two s3 files can take a prohibitively long time. "
+                "Approximate access time (using `.load()`) can exceed 6 minutes per data "
+                "variable.",
+                stacklevel=2,
+            )
+            _confirm_proceed()
```

Comment:
@rwegener2 I wanted to trigger the `_confirm_proceed()` so I passed in a list of s3 urls... I got the confirmation, but the warning never showed (and the confirmation is confusing without it). Wondering if it makes sense to kick the warning into the function?

Reply:
Hm, interesting. The warning appeared when I was running it. That was a few commits back, though, so 🤷‍♀️ 
I think it's fine to add the warning to the `_confirm_proceed()` function. My guess is that either way we'll need to increase the `stacklevel` to 3 or 4 if the warning isn't showing up at level 2.

Reply:
I seem to recall I was messing with the warnings settings in my hub for a presentation recently... since this is an important one, I added a cell to turn warnings on (changing the stack level didn't help in my case). Let me know if you think we should take a different approach. Otherwise, this should be good to go.